### PR TITLE
Change security flags

### PR DESCRIPTION
**Pull request by mberhault**
Addresses #4269

Replace `--certs` with individual flags for CA cert and key, and
server/client cert and key.
One notable difference is that we now have a single certificate for
nodes with double role as server and client authentication.

A few notes:
- this turned out to be big enough, so I'll leave the default `--insecure` value to the next PR.
- removed cmd/monitor, it isn't being used anywhere
- I'll probably collapse all the path building into helder funcs.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4957)

### CODE REVIEW

**Comment by petermattis** [comment]
Looks good besides the `{ssl,SSL}` prefix. @spencerkimball, @bdarnell might want to weigh in on the flag naming. 

---

Review status: 0 of 43 files reviewed at latest revision, 2 unresolved discussions.

---

*[cli/flags.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIcSDOZH5fQqVdTEgv:-KCIcSDOZH5fQqVdTEgw:2048723894) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/flags.go#L153)):*
Looks like these flag names are inspired by postgres. I'm wondering if we want the `ssl-` prefix here. We're using TLS, not SSL. Postgres likely uses `ssl` for historical reasons. But newer programs like `docker` use `tls` for a prefix. `etcd` omits a prefix entirely and uses `--ca-file` and `--cert-file`.

---

*[security/x509.go, line 151 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIdXFtxz7tJ_lJHYC7:-KCIdXFtxz7tJ_lJHYC8:1380754966) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/security/x509.go#L151)):*
What does this change do?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-:-KCIdd-ocDNBhkwdDrfV:-2135749503)*

File: cli/flags.go [hunk 1]
```diff
@@ -145,6 +141,18 @@ duration.`),
 Adjusts the max idle time of the scanner. This speeds up the scanner on small
 clusters to be more responsive.`),
 
+	"ssl-ca": wrapText(`
```
Comment (spencerkimball) [inline cli/flags.go]: Why don't we rename these keys so they're shorter and more consistent:

```
--ca-cert
--ca-key
--cert
--key
```

File: cli/kv.go [hunk 2]
```diff
@@ -37,11 +37,13 @@ func makeDBClient() (*client.DB, *stop.Stopper) {
 	// TODO(marc): KV endpoints are now restricted to node users.
 	// This should probably be made more explicit.
 	db, err := client.Open(stopper, fmt.Sprintf(
-		"%s://%s@%s?certs=%s",
+		"%s://%s@%s?sslca=%s&sslcert=%s&sslcertkey=%s",
```
Comment (spencerkimball) [inline cli/kv.go]: same with these... `?ca=%s&cert=%s&key=%s`

File: cli/flags.go [hunk 3]
```diff
@@ -145,6 +141,18 @@ duration.`),
...
+Path to the CA certificate. Needed by servers and clients in secure mode.`),
```
Comment (spencerkimball) [inline cli/flags.go]: s/servers and clients/clients and servers/  (for consistency with how you phrase them elsewhere).

**Comment by bdarnell** [comment]
Review status: 0 of 43 files reviewed at latest revision, 6 unresolved discussions.

---

*[cli/flags.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIcSDOZH5fQqVdTEgv:-KCIiFaG7dJiX46qFn99:256065237) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/flags.go#L153)):*
Users will still need to see names with `ssl`: the postgres connection url uses parameters like `sslcert` and `sslkey`. I think it's probably better to just use those names (without hyphens) than to use different names between our own tools and postgres drivers (in fact, we should have the option to just pass in a postgres url for all of our client-side commands). If we're going to deviate from postgresql's naming then I'd prefer to drop the tls/ssl prefix entirely (and `cert-key` is weird; it's just a key)

---

*[resource/test_certs/README.md, line 23 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIkMoME0xHxwzhPzyP:-KCIkMoNBA-kf2r0Ti9e:743052973) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/resource/test_certs/README.md#L23)):*
Specifying these separately is really tedious. I get that we want to  move away from our old assumption that everything is in one place (especially for the CA key, which really shouldn't be copied all over the place), but I think it would be good if we could still infer the locations of the other files from the location of one.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957)*

**Comment by tamird** [comment]
Reviewed 42 of 43 files at r1.
Review status: 42 of 43 files reviewed at latest revision, 8 unresolved discussions.

---

*[acceptance/cluster/localcluster.go, line 398 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIvpaWfQpcrVg-uXl2:-KCIvpaWfQpcrVg-uXl3:-756917496) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/acceptance/cluster/localcluster.go#L398)):*
I think this function (and perhaps the others) should take a struct as their argument, so that the fields are more clearly defined in the caller.

---

*[base/context.go, line 55 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIw4Xj3sBVgIZ1Bo8s:-KCIw4Xkc_mZAO0XG1ZK:-1855018278) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/base/context.go#L55)):*
this could use the same struct described in my other comment

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957)*

**Comment by petermattis** [comment]
Review status: 42 of 43 files reviewed at latest revision, 8 unresolved discussions.

---

*[cli/flags.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIcSDOZH5fQqVdTEgv:-KCLPyXZJSYC0MUhESD4:315880750) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/flags.go#L153)):*
Yes, the postgres connection url uses parameters with the `ssl` prefix, but I'm not sure if that is enough justification to use the `ssl` prefix for our flags. My preference is to try the prefix entirely and go with `--{,ca-}{cert,key}`. Sounds like this is @spencerkimball's preference as well. Let's finalize this decision today.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957)*

**Comment by spencerkimball** [comment]
Review status: 42 of 43 files reviewed at latest revision, 8 unresolved discussions.

---

*[cli/flags.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIcSDOZH5fQqVdTEgv:-KCLRB1T5orQnNqAGWVG:-1793157539) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/flags.go#L153)):*
I agree we should be able to accept a URL on the `sql` command!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957)*

**Comment by petermattis** [comment]
Review status: 42 of 43 files reviewed at latest revision, 8 unresolved discussions.

---

*[cli/flags.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIcSDOZH5fQqVdTEgv:-KCLRmMIXtlaec_z8OpI:442709246) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/flags.go#L153)):*
`cockroach start --url <url>` already works. Is there something more you're looking for.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957)*

**Comment by petermattis** [comment]
Review status: 42 of 43 files reviewed at latest revision, 8 unresolved discussions.

---

*[cli/flags.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIcSDOZH5fQqVdTEgv:-KCLS1bZn2S0NeIgWni9:1352229759) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/flags.go#L153)):*
Err, `cockroach sql --url <url>`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957)*

**Comment by bdarnell** [comment]
Review status: 42 of 43 files reviewed at latest revision, 8 unresolved discussions.

---

*[cli/flags.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIcSDOZH5fQqVdTEgv:-KCL_yqkr9Jdg0LkT3Y8:-1848109763) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/flags.go#L153)):*
Ah, I missed that we already had that. I would like to take it a step further, though, and use it for *all* client commands, including those that are based on the KV api under the hood (unless all of those are doomed to go away or be replaced by sql)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957)*

**Comment by petermattis** [comment]
Review status: 42 of 43 files reviewed at latest revision, 8 unresolved discussions.

---

*[cli/flags.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIcSDOZH5fQqVdTEgv:-KCM3WPforbWRSV2lbWL:621498408) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/flags.go#L153)):*
The `--url` flag also works with the `cockroach {zone,user}` commands. I suppose we can make the `{range,node,kv,exterminate,quit}` commands also accept `--url`. File a bug. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957)*

**Comment by mberhault** [comment]
Review status: 42 of 43 files reviewed at latest revision, 8 unresolved discussions.

---

*[acceptance/cluster/localcluster.go, line 398 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIvpaWfQpcrVg-uXl2:-KCMe-IA8ud7VsBtJKc7:1628619989) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/acceptance/cluster/localcluster.go#L398)):*
it's just a function call, putting all args into a struct doesn't particularly make it clearer.

---

*[base/context.go, line 55 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIw4Xj3sBVgIZ1Bo8s:-KCMeIoQYTYL60OgZlcg:1565490536) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/base/context.go#L55)):*
ditto, adding an extra level of addressing seems odd.

---

*[cli/flags.go, line 144 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIHAvnXSTYxhGnoINj-r1-144:-KCMeGFiMeMnnSOAroEM:-356545086) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/flags.go#L144)):*
discussing in main thread about this

---

*[cli/flags.go, line 145 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIHAvnXSTYxhGnoINj-r1-145:-KCMeF5oztiPlKQMmaE3:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/flags.go#L145)):*
Done.

---

*[cli/flags.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIcSDOZH5fQqVdTEgv:-KCMeNmYtSbqIxByzfer:-1936021196) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/flags.go#L153)):*
about `--url`: it's fine for sql addresses, and to some extent for the `kv` layer comands as well, although there is some question of what to do with the security mode. use `sslmode` or the `scheme` part of the url?
This still leaves the start and certs commands which need actual flags (URLs don't make sense here).

I like using `ssl` to make it clear what cert/key we're talking about. Having simply `key` or `ca` is too ambiguous.
As for `tls`, I don't think it makes more sense. people equate `ssl` with secure communications and certs, often using the term as a superset of both actual `ssl` and `tls`.

Overall, it would be nice to have agreement between url args and cmdline args, but I really don't like the non-hyphened versions in postgres, and `sslrootcert` is also a terrible name.

So here are a few proposals:
1) stick with postgres names (`sslrootcert`, `sslcert`, `sslkey`) and add `sslrootkey` when signing certs.
2) customize names: `ssl-ca-cert`, `ssl-ca-key`, `ssl-cert`, `ssl-key`
3) custom, but dropping ssl entirely: `ca-cert`, `ca-key`, `cert`, `key`.

---

*[cli/kv.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIHAvnXSTYxhGnoINk-r1-40:-KCMgBVH--ObdDn6gBZF:-1343279657) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/kv.go#L40)):*
these should end up matching whatever we decide as the final names.

---

*[resource/test_certs/README.md, line 23 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIkMoME0xHxwzhPzyP:-KCMgO5hXjQnYJRQcYqg:396551727) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/resource/test_certs/README.md#L23)):*
I agree, but the old "custom directory + hard-coded filenames" was really odd and inflexible.
I'm not sure picking something in-between would be much better. We could either allow both directory and custom filenames, or deduce the directory from one of the filenames but the question becomes which.

---

*[security/x509.go, line 151 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIdXFtxz7tJ_lJHYC7:-KCMgEmBo8wubTd1Vkjj:-139168932) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/security/x509.go#L151)):*
this allows the node certificate to be used both as server and client certificates.
before this change, we generated two separate ones, but we can combine them.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957)*

**Comment by tamird** [comment]
Review status: 42 of 43 files reviewed at latest revision, 7 unresolved discussions.

---

*[acceptance/cluster/localcluster.go, line 398 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIvpaWfQpcrVg-uXl2:-KCMhvrqZUS8ce86T9B-:164446797) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/acceptance/cluster/localcluster.go#L398)):*
What do you mean? If you made it a struct it would be
```
security.RunCreateNodeCert(security.CreateNodeCertConfig{
  CertPath: ..,
  CAPath: ...,
 ....
})
```

Seems much clearer to me.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957)*

**Comment by mberhault** [comment]
per discussion, renamed to `ca-cert`, `ca-key`, `cert`, and `key`

---

Review status: 38 of 43 files reviewed at latest revision, 7 unresolved discussions.

---

*[acceptance/cluster/localcluster.go, line 398 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIvpaWfQpcrVg-uXl2:-KCN6o7hR2XLggxoKiSS:-2118918421) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/acceptance/cluster/localcluster.go#L398)):*
this doesn't really buy anything. with decent indentation and clear filenames as we have here, it's clear enough. Creating a new struct for each of those functions really won't help.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-:-KCN6hmv652h-ubvQLiv:-457447446)*

**Comment by tamird** [comment]
Reviewed 4 of 5 files at r2.
Review status: 42 of 43 files reviewed at latest revision, 7 unresolved discussions.

---

*[acceptance/cluster/localcluster.go, line 398 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIvpaWfQpcrVg-uXl2:-KCN9HIESiiyvYsolORj:-1501290253) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/acceptance/cluster/localcluster.go#L398)):*
OK. I disagree, but we don't have guidelines on this. Flagging for the style guide @bdarnell @petermattis 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957)*

**Comment by bdarnell** [comment]
LGTM

---

Review status: 34 of 43 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[acceptance/cluster/localcluster.go, line 398 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIvpaWfQpcrVg-uXl2:-KCRUTR4Lx0fIXqvvhuL:-1233086813) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/acceptance/cluster/localcluster.go#L398)):*
-1 on using a struct for this. It's not clear to me that it's even an improvement (The struct makes it possible to name your arguments but it also means that they are all optional as far as the compiler is concerned.), and even if it were an improvement it's such a minor thing that it wouldn't make sense to mandate in the style guide.

---

*[cli/cli_test.go, line 169 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCRV3JdvvhFGl4Bq_f1:-KCRV3Je32T1ot_vY6M-:-1808535745) ([raw file](https://github.com/cockroachdb/cockroach/blob/96590c45855dbc1cfa33dcf09ab3f0dda15d794c/cli/cli_test.go#L169)):*
Isn't `--insecure=false` implied by the presence of the cert/key flags?

---

*[cli/flags.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIcSDOZH5fQqVdTEgv:-KCRVEHRi7HC6PEVeLMu:-164911677) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/flags.go#L153)):*
The `scheme` part of the url is always `postgresql` and doesn't indicate security; the security mode would always come from `sslmode` when a url is used. 

I think I like the bare names (option 3) best, followed by postgres-style (option 1), prefixed with `tls-` (not listed), and prefixed by `ssl-` (option 2)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-:-KCRX7YINolLTeu_xnBS:676834233)*

**Comment by tamird** [comment]
Reviewed 8 of 9 files at r3.
Review status: 42 of 43 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957)*

**Comment by petermattis** [comment]
LGTM

---

Review status: 42 of 43 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[base/context.go, line 54 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCR_UQq7wv1t85QV9Yh:-KCR_UQq7wv1t85QV9Yi:1391255626) ([raw file](https://github.com/cockroachdb/cockroach/blob/96590c45855dbc1cfa33dcf09ab3f0dda15d794c/base/context.go#L54)):*
Should the `SSL` prefix be removed from these fields? I see utility in having the prefix, but `SSLCA` is quite a mouthful.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-:-KCRa9HbTmqysb6qvvwV:676834233)*

**Comment by mberhault** [comment]
Review status: 28 of 42 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[base/context.go, line 54 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCR_UQq7wv1t85QV9Yh:-KCXBNdnmgKslYt8bJBJ:-1384905262) ([raw file](https://github.com/cockroachdb/cockroach/blob/96590c45855dbc1cfa33dcf09ab3f0dda15d794c/base/context.go#L54)):*
This is a case where I'd much rather be very clear. SSLCA may be annoying from a capitalization point of view, but I think the clearer meaning in callers makes it worth it.

---

*[cli/cli_test.go, line 169 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCRV3JdvvhFGl4Bq_f1:-KCXBa_ddOGovgt4i3N6:1641726538) ([raw file](https://github.com/cockroachdb/cockroach/blob/96590c45855dbc1cfa33dcf09ab3f0dda15d794c/cli/cli_test.go#L169)):*
nope. Right now, specifying `--insecure=false` and empty `--cert` on the server is an error. 
On the client, it gets turned into `InsecureSkipVerify` mode.

---

*[cli/flags.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957#-KCIcSDOZH5fQqVdTEgv:-KCXBkuQVQjeXtAEXw_O:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/97f22ae40b697067e84b855c1ec1b11189ea6c76/cli/flags.go#L153)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4957)*

**Comment by jseldess** [comment]
Docs updated with https://github.com/cockroachdb/docs/pull/117.

**Pull request merged**